pasha/functor.py

```diff
@@ -286,7 +286,16 @@ def iterate(self, share):
             for f in subobj.files:
                 f.close()
 
-        it = zip(range(*share.indices(self.n_trains)), subobj.trains())
+        from extra_data import KeyData
+        if isinstance(subobj, KeyData):
+            # KeyData objects would not include an empty result by
+            # default, which may cause the index to not be aligned to
+            # the train ID list.
+            train_it = subobj.trains(include_empty=true)
```

Comment:
```suggestion
            train_it = subobj.trains(include_empty=True)
```

Also, note that by default it will use `None` as the value for empty trains. You can get a zero length array by passing `keep_dims=True` as well, but that will also change the shape of the arrays with data (except for xtdf detectors).

Reply:
Oops, how did I manage to lowercase it before committing.

But yes, that would be the expected behaviour. Frankly I'd like to somehow enable passing arguments to functors at some point, but for now `None` seems like a fine default. Lots of functions will need to end up checking anyway (and it's easy to check for `None`), unless you don't want empty trains and drop them accordingly before.